xc7/fasm2bels/pss_models.py

```diff
@@ -0,0 +1,938 @@
+            (node_wire_pkey, )
+        )
+        phy_tile_pkey, wire_in_tile_pkey = c.fetchone()
+        upstream_sink_wire_in_tile_pkey = check_for_default(wire_in_tile_pkey)
+
+        # Got a PPIP, move upstream.
+        if upstream_sink_wire_in_tile_pkey is not None:
+            c.execute(
+                "SELECT pkey FROM wire WHERE wire_in_tile_pkey = ? AND phy_tile_pkey = ?;",
+                (
+                    upstream_sink_wire_in_tile_pkey,
+                    phy_tile_pkey,
+                )
+            )
+            upstream_sink_wire_pkey = c.fetchone()[0]
+
+            if DEBUG:
+                print(upstream_sink_wire_pkey, "", end='')
+
+            # Recurse
+            return expand_ps_sink(
+                conn=conn,
+                c=c,
+                sink_wire_pkey=upstream_sink_wire_pkey,
+                source_to_sink_pip_map=source_to_sink_pip_map,
+                check_for_default=check_for_default
+            )
+
+    # No active PIP and PPIP, the sink is unconnected.
+    return None
+
+
+def check_for_active_ps_sinks(conn, db, active_pips, tile, site):
+    """
+    Loops over all PS7 inpus pins and checks if any of them is connected to
+    the PL through an active PIP. Returns True when at least one is, False
+    otherwise.
+    """
+
+    if DEBUG:
+        print("Checking PS7 active sinks...")
+
+    # Build source to sink map
+    source_to_sink_pip_map = {}
+
+    for sink_wire_pkey, source_wire_pkey in active_pips:
+        assert source_wire_pkey not in source_to_sink_pip_map
+        source_to_sink_pip_map[source_wire_pkey] = sink_wire_pkey
+
+    # Create PPIP checker.
+    check_for_default = create_check_for_default(db, conn)
+
+    # Identify sinks
+    sinks = []
+    for signal, width in PS7_PINS["input"]:
+        if width == 1:
+            sinks.append(signal)
+        else:
+            for i in range(width):
+                sinks.append("{}{}".format(signal, i))
+
+    prefix = site.type
+    wires = [prefix + "_" + s for s in sinks]
+
+    # Expand each wire sink upstream until an active PIP is found.
+    cur = conn.cursor()
+    wire_pkeys = [get_wire_pkey(conn, tile, w) for w in wires]
+    for wire, wire_pkey in zip(wires, wire_pkeys):
+
+        if DEBUG:
+            print("//", wire, "", end='')
+
+        # Check if it is connected to anything. If it is so then return True,
+        # there is no need to check further.
+        pkey = expand_ps_sink(
+            conn=conn,
+            c=cur,
+            sink_wire_pkey=wire_pkey,
+            source_to_sink_pip_map=source_to_sink_pip_map,
+            check_for_default=check_for_default
+        )
+
+        if DEBUG:
+            print(pkey)
+
+        if pkey is not None:
+            return True
+
+    # None of PS sinks is connected.
+    return False
+
+
+# =============================================================================
+
+
+def expand_ps_source(
```

Comment:
Rather than doing it this way, can you use the cleanup strategy seen here: https://github.com/SymbiFlow/symbiflow-arch-defs/blob/master/xc7/fasm2bels/clb_models.py#L240

Reply:
Basically the CARRY4 is similiar to the PS7 in that it is possible for a CARRY4 to be used soley for the COUT, which doesn't set any bitstream bits.  So the clb_models.py always emits it, and if none of the sources or sinks are used, it is deleted.  This strategy avoids needing model how the PS7 is connected.

Reply:
I agree that this can be done according to the cleanup strategy. But still the only way to determine whether the PS7 is in use it to check activity of particular PIPs. And the thing is that when I emit the PS7 and it is not used then `make_routes` complains about orphaned sinks even before the cleanup. But off course, I can retry that approach.

Reply:
We should probably add something to the PS7 sources that is something akin to "source is allowed to not sink".  I believe that is more general than this approach.

Reply:
@litghost I added the PS7 cleanup function and made `make_routes` not complain about unrouted sources/sinks for PS7.